app/src/main/java/io/zenandroid/onlinego/ui/screens/stats/StatsFragment.kt

```diff
@@ -34,5 +73,301 @@ class StatsFragment : Fragment(), StatsContract.View {
+            if(lastAddedTimestamp.isEmpty() || (lastAddedTimestamp.toInt() - currentTimestamp.toInt() > 60*60*24)) {
+                //added in hours - we don't need more precision on the line chart points
+                rankGraphEntries.add(Entry(currentTimestamp.toInt() / 60 / 60f, currentRank))
+                lastAddedTimestamp = currentTimestamp
+            }
+        }
+
+        if (rankGraphEntries.isNotEmpty()) {
+            Collections.sort(rankGraphEntries, EntryXComparator())
+            val rankDataSet = LineDataSet(rankGraphEntries, "Games")
+            rankDataSet.setDrawCircles(false)
+            rankDataSet.setDrawValues(false)
+            rankDataSet.lineWidth = 1.5f
+            rankDataSet.color = resources.getColor(R.color.rankGraphLine)
+            rankDataSet.mode = LineDataSet.Mode.HORIZONTAL_BEZIER;
+            val lineData = LineData(rankDataSet)
+            rankGraph.data = lineData
+        }
+        rankGraph.xAxis.position = XAxis.XAxisPosition.BOTTOM;
+        rankGraph.xAxis.valueFormatter = DayAxisValueFormatter(rankGraph)
+
+        rankGraph.xAxis.setLabelCount(6, true)
+        rankGraph.xAxis.textColor = resources.getColor(R.color.colorText)
+        rankGraph.xAxis.setDrawGridLines(false)
+        rankGraph.axisLeft.setDrawGridLines(false)
+        rankGraph.axisLeft.textColor = resources.getColor(R.color.colorText)
+        rankGraph.axisRight.setDrawGridLines(false)
+        rankGraph.axisRight.valueFormatter = RankFormatter()
+        rankGraph.axisRight.textColor = resources.getColor(R.color.colorText)
+        rankGraph.description.isEnabled = false
+        rankGraph.setDrawMarkers(false)
+        rankGraph.setNoDataText("No ranked games on record")
+        rankGraph.setNoDataTextColor(resources.getColor(R.color.colorActionableText))
+        rankGraph.legend.isEnabled = false
+        rankGraph.invalidate()
+    }
+
+    private fun fillOutcomePieChart(gameStats: MutableList<MutableMap<String, String>>) {
+        val outcomes = gameStats.groupingBy{ it["outcome"] }.eachCount()
+        if (outcomes.isNotEmpty()) {
+            val lostEntry = PieEntry((outcomes["0"] ?: 0).toFloat(), "Lost")
+            val wonEntry = PieEntry((outcomes["1"] ?: 0).toFloat(), "Won")
+            val entries = mutableListOf(lostEntry, wonEntry)
+            val dataSet = PieDataSet(entries, "Games")
+            dataSet.setColors(intArrayOf(R.color.chartLost, R.color.chartWon), context)
+            dataSet.setDrawValues(false)
+            val pieData = PieData(dataSet)
+            chart.data = pieData
+            val gamesWon = wonEntry.value.toInt();
+            val gamesLost = lostEntry.value.toInt()
+            val gamesWonPercent = (gamesWon * 100 / (gamesWon + gamesLost).toFloat())
+            val gamesLostPercent = 100 - gamesWonPercent
+
+            val centerText = SpannableString("${gamesWon}  ${String.format("%.1f", gamesWonPercent)}%\n" +
```

Comment:
I am not familiar with the graph API, but can't this be done without `Spannables`? I am having a lot of trouble understanding the code here...

Reply:
the piechart only has the one center text field, so the spannables were needed to achieve the multi-line/multi-color display

Reply:
Indeed the spannables are the easiest way to put this there. I also discovered there's a KTX DSL for writing nicer spannables, without using indexes. See the *second* answer here: https://stackoverflow.com/questions/10828182/spannablestringbuilder-to-create-string-with-multiple-fonts-text-sizes-etc-examp/23400573

Reply:
Fixed in https://github.com/acristescu/OnlineGo/commit/1c57f6a852393483b9a7b3936e3048083c23aed1